test/get/enforce_lockfile_test.dart

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:pub/src/exit_codes.dart';
+import 'package:test/test.dart';
+
+import '../descriptor.dart';
+import '../test_pub.dart';
+
+Future<void> main() async {
+  test('Recreates .dart_tool/package_config.json, redownloads archives',
+      () async {
+    final server = await servePackages();
+    server.serve('foo', '1.0.0');
+    await appDir({'foo': 'any'}).create();
+    await pubGet();
+    final packageConfig =
+        File(path(p.join(appPath, '.dart_tool', 'package_config.json')));
+    packageConfig.deleteSync();
+    await runPub(args: ['cache', 'clean', '-f']);
+    await pubGet(args: ['--enforce-lockfile']);
+    expect(packageConfig.existsSync(), isTrue);
+    await cacheDir({'foo': '1.0.0'}).validate();
+    await appPackageConfigFile([
+      packageConfigEntry(name: 'foo', version: '1.0.0'),
+    ]).validate();
+  });
+
+  test('Refuses to get if no lockfile exists', () async {
+    await appDir({}).create();
+    await pubGet(
+        args: ['--enforce-lockfile'],
+        error:
+            'Retrieving dependencies failed. Cannot do `--enforce-lockfile` without an existing `pubspec.lock`.');
+  });
+
+  test('Refuses to get if lockfile is missing package', () async {
+    final server = await servePackages();
+    server.serve('foo', '1.0.0');
+    await appDir({}).create();
+    await pubGet();
+    await appDir({'foo': 'any'}).create();
+
+    await pubGet(
+        args: ['--enforce-lockfile'],
+        error: 'Dependency foo is not already locked in `pubspec.lock`.');
+  });
+
+  test('Refuses to get if package is locked to version not matching constraint',
```

Comment:
Add edge case where `1.0.0` was removed from the server, but version `1.1.0` is still available, and `1.0.0` is in pubspec.lock